sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -52,6 +52,68 @@ object InMemoryRelation {
 private[columnar]
 case class CachedBatch(numRows: Int, buffers: Array[Array[Byte]], stats: InternalRow)
 
+private[columnar] class CachedBatchIterator(
+    rowIterator: Iterator[InternalRow],
+    output: Seq[Attribute],
+    batchSize: Int,
+    useCompression: Boolean,
+    batchStats: LongAccumulator,
+    singleBatchPerPartition: Boolean) extends Iterator[CachedBatch] {
+
+  def next(): CachedBatch = {
+    val columnBuilders = output.map { attribute =>
+      ColumnBuilder(attribute.dataType, batchSize, attribute.name, useCompression)
+    }.toArray
+
+    var rowCount = 0
+    var totalSize = 0L
+
+    val terminateLoop = (singleBatch: Boolean, rowIter: Iterator[InternalRow],
```

Comment:
@CodingCat Could you explain me what `singleBatch` means here? I cannot get my head around it:) Thanks!

Reply:
to make getting partition stats easier, we construct only one CatchedBatch for each partition when enabling the functionality proposed in this PR. `singleBatch` distinguishes the scenarios which enables/disables the functionality by introducing different while loop termination conditions, making the other code reusable